### ISSUE TITLE
OHEM solution? Train with true negative images?

### ISSUE THREAD

**Issue by yanhn** [zhreshold/mxnet-ssd#163]
I want to do hard negative mining with true negative images.
Any solutions?
I found the `multiboxtarget.cc` has parameters for negative_threshold and negative_ratio, but it only deal with negative proposal within the images(those with ground truth label). I want to add true negative images to trainning.
My current idea is to feed true negative images to the net.
I've generated `train.lst` using `prepare_dataset.py` with flag `--true-negative True`, but encountered an error during `save_imglist`.
Seems like the program skip this situation when label doesn't have any object according to line 76 in imdb.py.`if label.size < 1: continue`
So how can I do this job. Any idea would be appreciated, thanks.

**Comment by yanhn** [comment]
I generated empty xml for true nagative images and put them all in trainSet.
My `train.lst` look like this:
![image](https://user-images.githubusercontent.com/12744658/34559496-06afecbe-f17d-11e7-8e68-43ad7c87909d.png)
and resulted in an error:
`src/io/image_det_aug_default.cc:258: Check failed: label_width >= 7 (2 vs. 7)` which lead me to 
![image](https://user-images.githubusercontent.com/12744658/34559579-4c22f35e-f17d-11e7-8235-97c552034ca3.png)

**Comment by zhreshold** [comment]
You can try negative image padded with -1 labels.
Make sure you have positive image in each batch otherwise it’s illegal for training

**Comment by yanhn** [comment]
Thanks @zhreshold , the training works for now.
My data batch looks like this:
![image](https://user-images.githubusercontent.com/12744658/34610178-ef4b46b6-f25a-11e7-90cd-d6b20f522752.png)
In this way the `multiboxtarget` layer can do the hard mining when I feed a image with labels all set to -1?
I'll also do some experiments after some epoch.

**Comment by yanhn** [comment]
I did some experiments and found the bad case didn't improve apparently.
I followed your advice to pad the negative images' label with -1 and avoid all neg batch. The data batch looks like the pic in my last comments. But the results are not satisfying. So is there something I missed? Thanks.

**Comment by zhreshold** [comment]
It's normal you don't get any improvement using pure negative images.
The reason is simple, providing lots of background training data won't benefit the detector too much, unless your training data is very limited, because the information provided by them are very limited(no class info, no object shape,  etc), only beneficial to determine if an area is object/background.

**Comment by yanhn** [comment]
In my opinion, I can feed some image close to my target but actually not the class. And let the model learn the minor difference, so the benefit is hard to tell.
And I found in `multibox_target.cc` that if I set the true negative images' labels to "-1", it will do nothing but skip this true negative images.  Cos the num_valid_gt remains zero. This maybe the reason for my issue.
```
    int num_valid_gt = 0;
    for (int i = 0; i < num_labels; ++i) {												//label iter to calcu valid gt
      if (static_cast<float>(*(p_label + i * label_width)) == -1.0f) {					//every iter consists 5 num?????????????
        CHECK_EQ(static_cast<float>(*(p_label + i * label_width + 1)), -1.0f);
        CHECK_EQ(static_cast<float>(*(p_label + i * label_width + 2)), -1.0f);
        CHECK_EQ(static_cast<float>(*(p_label + i * label_width + 3)), -1.0f);
        CHECK_EQ(static_cast<float>(*(p_label + i * label_width + 4)), -1.0f);
        break;
      }
      ++num_valid_gt;
    }  // end iterate labels
    if (num_valid_gt > 0) { ...} 
```

**Comment by yanhn** [comment]
And I think something strange with line220 & line225 of `multibox_target.cc`, for k is the class idx and j is the anchor idx.
`DType tmp = p_cls_preds[j + num_anchors * k];`

**Comment by zhreshold** [comment]
You can change the behavior if you like.

**Comment by yanhn** [comment]
My bad about j & k question. Nothing's strange about the code.
I am now back to find true hard negative solution. Still trapped.
